presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergResourceFactory.java

```diff
@@ -0,0 +1,113 @@
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.security.ConnectorIdentity;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+import org.apache.hadoop.fs.Path;
+import org.apache.iceberg.CatalogUtil;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.SupportsNamespaces;
+
+import javax.inject.Inject;
+
+import java.util.concurrent.ExecutionException;
+
+import static com.facebook.presto.spi.StandardErrorCode.NOT_SUPPORTED;
+import static com.google.common.base.Throwables.throwIfInstanceOf;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static java.util.Objects.requireNonNull;
+import static org.apache.iceberg.CatalogProperties.WAREHOUSE_LOCATION;
+
+/**
+ * Factory for loading Iceberg resources such as Catalog and FileIO.
+ */
+public class IcebergResourceFactory
+{
+    private final Cache<String, Catalog> catalogCache;
+
+    private final String catalogName;
+    private final String catalogImpl;
+    private final String catalogWarehouse;
+    private final HdfsEnvironment hdfsEnvironment;
+
+    @Inject
+    public IcebergResourceFactory(IcebergConfig config, IcebergCatalogName catalogName, HdfsEnvironment hdfsEnvironment)
+    {
+        this.catalogName = requireNonNull(catalogName, "catalogName is null").getCatalogName();
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsConfiguration");
+        requireNonNull(config, "config is null");
+        this.catalogImpl = config.getCatalogImpl();
+        this.catalogWarehouse = config.getCatalogWarehouse();
+        catalogCache = CacheBuilder.newBuilder()
+                .maximumSize(config.getCatalogCacheSize())
+                .build();
+    }
+
+    public Catalog getCatalog(ConnectorSession session)
+    {
+        try {
+            return catalogCache.get(getCatalogCacheKey(session), () -> CatalogUtil.loadCatalog(catalogImpl, catalogName,
+                    ImmutableMap.of(WAREHOUSE_LOCATION, catalogWarehouse),
+                    hdfsEnvironment.getConfiguration(new HdfsContext(session), new Path(catalogWarehouse))));
+        }
+        catch (ExecutionException | UncheckedExecutionException e) {
+            throwIfInstanceOf(e.getCause(), PrestoException.class);
+            throwIfUnchecked(e);
+            throw new UncheckedExecutionException(e);
+        }
+    }
+
+    public SupportsNamespaces getNamespaces(ConnectorSession session)
+    {
+        Catalog catalog = getCatalog(session);
+        if (catalog instanceof SupportsNamespaces) {
+            return (SupportsNamespaces) catalog;
+        }
+        throw new PrestoException(NOT_SUPPORTED, "Iceberg catalog implementation " + catalogImpl + " does not support namespace operations");
+    }
+
+    private String getCatalogCacheKey(ConnectorSession session)
```

Comment:
This is basically to get a string representation of the user identity. As long as the identity does not change, the same catalog implementation can be used. This technically should not need any cache expiration, because the Catalog instance itself does not check access, but when any catalog action is called the access of the identity is checked, so the cache can still work when the access policy of the identity changes at server side.